crypto/ube/ube.c

```diff
@@ -0,0 +1,238 @@
+}
+
+// TODO
+// These two functions currently mocks the backend implementations of the fork
+// detection method and snapsafe detection method. They will be reimplemented in
+// a future change.
+static int get_snapsafe_generation_number_mocked(uint64_t *gn) {
+  *gn = 1;
+  if (testing_snapsafe_generation_number != 0) {
+    *gn = testing_snapsafe_generation_number;
+  }
+  return 1;
+}
+static uint64_t get_fork_generation_number_mocked(void) {
+  if (testing_fork_generation_number != 0) {
+    return testing_fork_generation_number;
+  }
+  return 1;
+}
+
+
+// The UBE generation number is shared for the entire address space. One could
+// implement a per-thread UBE generation number. However, this could be
+// inefficient for global volatile states if care is not taken. Because every
+// thread only have visibility on their own per-thread generation number, we
+// could trigger a new randomization of the volatile state per-thread. While a
+// single randomization of the global volatile state should be sufficient.
+struct ube_global_state {
+  uint64_t generation_number;
+  uint64_t cached_fork_gn;
+  uint64_t cached_snapsafe_gn;
+};
+static struct ube_global_state ube_state = { 0 };
+
+// Convenience object that makes it easier to extend the number of detection
+// methods without having to modify function signatures.
+struct detection_gn {
+#define NUMBER_OF_DETECTION_GENERATION_NUMBERS 2
+  uint64_t current_fork_gn;
+  uint64_t current_snapsafe_gn;
+};
+
+// set_ube_detection_unavailable_once is the single mutation point of
+// |ube_detection_unavailable|. Sets the variable to 1 (true).
+static void set_ube_detection_unavailable_once(void) {
+  ube_detection_unavailable = 1;
+}
+
+// ube_failed is a convenience function to synchronize mutation of
+// |ube_detection_unavailable|. In practice, synchronization is not strictly
+// needed because currently the mutation is only ever assigning 1 to the
+// variable.
+static void ube_failed(void) {
+  CRYPTO_once(&ube_detection_unavailable_once, set_ube_detection_unavailable_once);
+}
+
+// ube_state_initialize initalizes the global state |ube_state|.
+static void ube_state_initialize(void) {
+
+  ube_state.generation_number = 0;
+  ube_state.cached_fork_gn = get_fork_generation_number_mocked();
+  int ret_snapsafe_gn =
+    get_snapsafe_generation_number_mocked(&(ube_state.cached_snapsafe_gn));
+
+  if (ube_state.cached_fork_gn == 0 ||
+      ret_snapsafe_gn == 0) {
+    ube_failed();
+  }
+}
+
+// ube_update_state updates the global state |ube_state| with the generation
+// numbers loaded into |current_detection_gn| and increments the UBE generation
+// number.
+static void ube_update_state(struct detection_gn *current_detection_gn) {
+
+  ube_state.generation_number += 1;
+
+  // Make sure we cache all new generation numbers. Otherwise, we might detect
+  // a fork UBE but, in fact, both a fork and snapsafe UBE occurred. Then next
+  // time we enter, a redundant reseed will be emitted.
```

Comment:
Can you add a test for this: both fork and snapsafe but ensure the generation number only increments once.

Reply:
That makes sense. I did that.